feat: add EmailPasswordSignUpForm component

HellWolf93 opened the pull request:
#### Changes proposed
- Add EmailPasswordSignUpForm component
- Add tests for EmailPasswordSignUpForm component
- Add docs for EmailPasswordSignUpForm component

LeandroTorresSicilia commented on cypress/integration/auth/emailPasswordSignUpForm.js:
```diff
@@ -0,0 +1,37 @@
+        cy.get(EMAIL_INPUT).type('test@email.com');
+        cy.get(PASSWORD_INPUT).type('TestPassword');
+        cy.get(BUTTON).click();
+        cy.wait('@firebaseSignUp');
```
We should do assertions that check for something that indicates the user is authenticated like the new URL or an element that exists when authenticated.
Ideally, we should start up the emulator and login and then make the assertion but I don't know if right now there is an emulator for auth.

HellWolf93 replied:
As for now, only Database, Firestore, Functions and Hosting are available for emulator. If we are going to test this it will be necesary to hit the real API, and handle the creation and deletion of users each time the tests runs, we could do that but I dont think it's not a good idea, given that we can trust firebase's ```createUserWithEmailAndPassword``` function will correctly authenticate the user when the sign up is correct

LeandroTorresSicilia commented on cypress/integration/auth/emailPasswordSignUpForm.js:
```diff
@@ -0,0 +1,37 @@
+        cy.get(EMAIL_INPUT).type('test@email.com');
+        cy.get(PASSWORD_INPUT).type('TestPassword');
+        cy.get(BUTTON).click();
+        cy.get(SPINNER_BACKDROP).should('exist');
```
In the case of unit test is a good practice make only one assertion for a test, but in the case of integrations we can make more than on assertion, then I think we can check for the spinner in the same test above

LeandroTorresSicilia commented on cypress/integration/auth/emailPasswordSignUpForm.js:
```diff
@@ -0,0 +1,37 @@
+        cy.get(BUTTON).click();
+        cy.get(SPINNER_BACKDROP).should('exist');
+    });
+});
```
I think we could add another test case for when user type a wrong email or a too short password and then the error message is showed

LeandroTorresSicilia commented on packages/auth/docs/components/EmailPasswordSignUpForm.story.mdx:
```diff
@@ -0,0 +1,32 @@
+
+This component HAVE TO be used under a properly configured `RainbowFirebaseApp` component.
+
+### Translatable texts
```
Internationalization FormattedMessage keys

LeandroTorresSicilia commented on packages/auth/docs/stories/emailPasswordSignUpForm.story.js:
```diff
@@ -0,0 +1,183 @@
+        maxWidth: '400px',
+    };
+    const logoStyle = { textAlign: 'center' };
+    const logoTextStyle = { fontSize: '1.8em' };
```
since these styles are constants them can be defined outside the component

LeandroTorresSicilia commented on packages/auth/docs/stories/emailPasswordSignUpForm.story.js:
```diff
@@ -0,0 +1,183 @@
+                <rect width="11.84" height="11.84" x=".032" y="27.954" fill="#1AD1A3" rx="2" />
+            </g>
+        </svg>
+    );
```
same this can be outside or put this in another file to avoid pollute the example

LeandroTorresSicilia commented on packages/auth/docs/stories/emailPasswordSignUpForm.story.js:
```diff
@@ -0,0 +1,183 @@
+                <EmailPasswordSignUpForm
+                    nameIcon={nameIcon}
+                    emailIcon={emailIcon}
+                    passwordIcon={passwordIcon}
```
since we don't define any props the icons should be part of the component for now

LeandroTorresSicilia commented on packages/auth/docs/stories/emailPasswordSignUpForm.story.js:
```diff
@@ -0,0 +1,183 @@
+            </WhenAuthenticated>
+        </RainbowFirebaseApp>
+    );
+};
```
There is a way we can see all the example code in the documentation?

<img width="1118" alt="Screen Shot 2020-05-16 at 1 08 15 PM" src="https://user-images.githubusercontent.com/24737097/82127027-4f43cd00-9776-11ea-837c-d5860a1ed189.png">

LeandroTorresSicilia commented on packages/auth/src/components/EmailPasswordSignUpForm/index.js:
```diff
@@ -0,0 +1,120 @@
+    const { className, style } = props;
+    const [name, setName] = useState('');
+    const [email, setEmail] = useState('');
+    const [password, setPassword] = useState('');
```
We should use redux-form which is setup in RainbowFirebaseApp like react-intl, then it should work out of the box.
Redux-form take care of the state of the form and also make the validation through the validate function.
Note that we should add test for the validate function and also add test cases for when errors show up.

LeandroTorresSicilia commented on packages/auth/src/components/EmailPasswordSignUpForm/__test__/emailPasswordSignUpForm.spec.js:
```diff
@@ -0,0 +1,50 @@
+        setTimeout(() => {
+            expect(actions.showAppMessage).toHaveBeenCalledTimes(1);
+            done();
+        }, 0);
```
can we test this with async await instead of using setTimeout?

LeandroTorresSicilia commented on packages/auth/src/components/EmailPasswordSignUpForm/index.js:
```diff
@@ -0,0 +1,120 @@
+    );
+    const buttonLabel = (
+        <FormattedMessage id="EmailPasswordSignUpForm.signup" defaultMessage="Create account" />
+    );
```
this const can be outside the component to avoid recreation of them in each rerender

LeandroTorresSicilia commented on packages/auth/src/components/EmailPasswordSignUpForm/index.js:
```diff
@@ -0,0 +1,120 @@
+
+    const buttonStyle = {
+        width: '100%',
+    };
```
let's use styled-components for styles to maintain separation of concerns

LeandroTorresSicilia commented on packages/auth/src/components/EmailPasswordSignUpForm/index.js:
```diff
@@ -0,0 +1,120 @@
+                onChange={(event) => setName(event.target.value)}
+                required
+            />
+            <br />
```
it is better to use styles instead of creating dom element to separate components

LeandroTorresSicilia commented on packages/auth/src/helpers/intl-enzyme-test-helper.js:
```diff
@@ -0,0 +1,23 @@
+import { IntlProvider } from 'react-intl';
```
Good job!!!
let's change the file name to `intlEnzyme`, we will be using camel case instead of dashes

LeandroTorresSicilia commented on packages/auth/src/components/EmailPasswordSignUpForm/index.js:
```diff
@@ -0,0 +1,120 @@
+    passwordIcon: undefined,
+};
+
+export default EmailPasswordSignUpForm;
```
we need to export this component in `packages/auth/src/index.js`

LeandroTorresSicilia commented on cypress/integration/auth/emailPasswordSignUpForm.js:
```diff
@@ -0,0 +1,37 @@
+const PASSWORD_INPUT = 'input[name="password"]';
+
+describe('EmailPasswordSignUpForm', () => {
+    beforeEach(() => {
```
can we do this in the before hook instead of in order to call it only once?

HellWolf93 replied:
it is necessary to do this in beforeEach beause the server config is restored to its default between tests

LeandroTorresSicilia commented:
<img width="999" alt="Screen Shot 2020-05-16 at 12 08 42 PM" src="https://user-images.githubusercontent.com/24737097/82151580-b074aa80-9821-11ea-8372-12bf1f11f329.png">
I think we should fix how icons look when the spinner shows up, also, since the icons are internals they should change its colors with the theme

HellWolf93 commented:
> <img alt="Screen Shot 2020-05-16 at 12 08 42 PM" width="999" src="https://user-images.githubusercontent.com/24737097/82151580-b074aa80-9821-11ea-8372-12bf1f11f329.png">
> 
> I think we should fix how icons look when the spinner shows up, also, since the icons are internals they should change its colors with the theme

Fixed, just added `z-index` to the spinner container.

reiniergs commented on packages/auth/src/components/EmailPasswordSignUpForm/index.js:
```diff
@@ -0,0 +1,105 @@
+        id: 'EmailPasswordSignUpForm.namePlaceholder',
+        defaultMessage: 'Enter your name',
+    });
+    const emailPlaceholder = intl.formatMessage({
```
can we move all these message components into a different file e.g. message.js then you can export then and that way the logic for I18n is isolated?

reiniergs replied:
The <FormatedMessage /> components above could be moved out of this file also

HellWolf93 replied:
> can we move all these message components into a different file e.g. message.js then you can export then and that way the logic for I18n is isolated?

we should not do that because `intl` is obtained with useIntl, and hooks can only be called within components, and the placeholder should be a plain string

reiniergs replied:
yeah, I hear you! ... I guess I was thinking about using `defineMessage` API. Why we need this is the question :) ... because moving forward I believe It will be nice to develop messages in a different file as a best practice, kinda separation of concern.

reiniergs replied:
by the way of example:

```js
import {defineMessages, defineMessage} from 'react-intl';

export const messages = defineMessages({
  greeting: {
    id: 'app.home.greeting',
    description: 'Message to greet the user.',
    defaultMessage: 'Hello, {name}!',
  },
});
export const msg = defineMessage({
  id: 'single'
  defaultMessage: 'single message',
  description: 'header'
})
```

then you can use the hook and do the formatting in the component

more info here: https://formatjs.io/docs/react-intl/api

HellWolf93 replied:
thanks for the clarification and the example, i think it is OK now
